synapse/storage/databases/main/pusher.py

```diff
@@ -388,6 +393,57 @@ def _delete_pushers(txn) -> int:
 
         return number_deleted
 
+    async def _remove_deleted_email_pushers(
+        self, progress: dict, batch_size: int
+    ) -> int:
+        """A background update that deletes all pushers for email addresses no longer
+        associated with a user.
+        """
+
+        last_pusher = progress.get("last_pusher", 0)
+
+        def _delete_pushers(txn) -> int:
+
+            sql = """
+                SELECT p.id, p.user_name, p.add_id, p.pushkey
+                FROM pushers AS p
+                    JOIN user_threepids AS t ON t.user_id=p.user_name
+                WHERE p.app_id = 'm.email'
+                    AND t.medium = 'email'
+                    AND t.address = p.pushkey
+                    AND p.id > ?
+                ORDER BY p.id ASC
+                LIMIT ?
```

Comment:
I may be missing something, but won't this only delete pushers that *are* associated with a user account; whereas we want to be doing the opposite?

Reply:
Eek I forgot to test that! At least I spelled `p.app_id` wrong too!

This query should actually do what we want:
```sql
SELECT p.id, p.user_name, p.app_id, p.pushkey
FROM pushers AS p
    LEFT JOIN user_threepids AS t 
        ON t.user_id=p.user_name
        AND t.medium = 'email'
        AND t.address = p.pushkey
WHERE t.user_id is NULL
    AND p.app_id = 'm.email'
    AND p.id > ?
ORDER BY p.id ASC
LIMIT ?
```